### PR TITLE
Remove jQuery

### PR DESCRIPTION
**Pull request by sorin-davidoi**
~~This is based on #535 for two reasons:~~
- ~~I assumed that #535 is likely to be (eventually) merged, so I can rebase this later against `master` in that case~~
- ~~Make it easier for reviewers to test both of them at the same time~~

~~Depending on which browsers we need to support (couldn't find them listed), we might need the following polyfills~~:
- ~~[Array.from](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from#Browser_compatibility) (no IE)~~
- ~~[String.trim](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/Trim#Browser_compatibility) (no IE 8)~~
- ~~[Element.closest](https://caniuse.com/#feat=element-closest) (no IE)~~
- ~~[Promise](https://caniuse.com/#feat=promises) (no IE)~~
- ~~[fetch](https://caniuse.com/#feat=fetch) (no IE)~~

~~I couldn't test the changes in `editor.js`.~~

Reduces JavaScript payload size from 229.88 KB to 197.91 KB (measured on `book-example`). Helps address #475.

### CODE REVIEW

**Comment by Michael-F-Bryan** [comment]
> Depending on which browsers we need to support (couldn't find them listed)

I don't believe we've got a proper list of supported browsers. My perception is that the vast majority of our users are going to be developers or computer-savvy people, and as a very broad rule those sorts of people tend to use a fairly up-to-date version of Chrome or Firefox. If we can support IE then that's awesome, but if it requires large amounts of effort or hacky band-aid solutions to do then I'll leave the decision up to the developer...

**Comment by sorin-davidoi** [comment]
Well, ideally we would only load the polyfills for the browsers that need it (feature-detection) - but that is non-trivial without a build system (e.g. webpack). A simpler solution would be to just https://polyfill.io/v2/docs/. What do you think?

**Comment by sorin-davidoi** [comment]
I see that crates.io uses Google Analytics. Assuming there is a significant overlap between the audiences, maybe @carols10cents could share some insights about browser usage to guide this decision?

**Comment by sorin-davidoi** [comment]
Actually, crates.io [does not support IE](https://github.com/rust-lang/crates.io/blob/master/config/targets.js).

**Comment by steveklabnik** [comment]
I came to this thread to post RFC 1985.

**Comment by sorin-davidoi** [comment]
Rebased against master.

File: src/theme/playpen_editor/editor.js [hunk 1]
```diff
@@ -2,10 +2,10 @@ window.editors = [];
 (function(editors) {
     if (typeof(ace) === 'undefined' || !ace) {
         return;
-    }
-
-    $(".editable").each(function() {
-        let editor = ace.edit(this);
+    
+    
+    Arom.from(document.querySelectorAll('.editable')).forEach(function(editable) {
```
Comment (projektir) [inline src/theme/playpen_editor/editor.js]: What is `Arom`? Is this supposed to be `Array`?

Comment (sorin-davidoi) [reply src/theme/playpen_editor/editor.js]: Yes, fixed 👍

**Comment by projektir** [comment]
Have you tested this code? Specifically, did you test that the Ace Editor boxes still work?

**Comment by sorin-davidoi** [comment]
I mentioned in the description that the changes in `editor.js` have not been tested (couldn't manage to trigger the codepath - it does not even load the file in the browser). What variables need to be toggled to enable the functionality?

**Comment by sorin-davidoi** [comment]
Apparently, I've missed https://github.com/rust-lang-nursery/mdBook/issues/521 - just tested and it works fine from what I can tell.

**Comment by projektir** [comment]
Yes, the Ace editor is not yet documented, which is my bad! Some settings need to be enabled for it to work.

Generally, I test by running mdBook on the Rust by Example book, since that's what Ace editor was added for originally in the first place, so it's all setup there. I've just tested it with your branch and it seems to be working fine. :blush:

**Comment by Michael-F-Bryan** [comment]
I just checked it out on my machine and I think there's an issue with the code expansion/hiding feature.

For example, on [this page](https://rust-lang-nursery.github.io/mdBook/format/rust.html), here's what it looks like collapsed (the default):

screenshot_2018-01-16_193604

Then hit run and you can see what it prints out:

screenshot_2018-01-16_193619

Expanded:

screenshot_2018-01-16_193627

---

This is what it's actually meant to look like when expanded:

screenshot_2018-01-16_193914

**Comment by sorin-davidoi** [comment]
@Michael-F-Bryan Thanks, fixed! Sidenote, I really hate how you never know whether you are operating on a single element or on an array in jQuery (in this case, I assumed single element, so only the first hidden line was toggled).

**Comment by projektir** [comment]
Ah, blast, I didn't check code expansion, good catch @Michael-F-Bryan.

**Comment by Michael-F-Bryan** [comment]
I don't think there were any other noticeable bugs/regression. @sorin-davidoi, do you think this is ready to merge?

**Comment by sorin-davidoi** [comment]
Yes, I would say so.

**Comment by sorin-davidoi** [comment]
Is jQuery part of the official `mdBook` API? I'm thinking that this (and possibly #550) may be breaking scripts people placed in `additional-js`.

**Comment by Michael-F-Bryan** [comment]
> Is jQuery part of the official mdBook API? 

I don't think we've ever said it's guaranteed to be loaded, it just happens to be because we use it internally. I'm going to say that people should make sure they explicitly load jQuery if they need it, plus this PR will be part of a major semver bump, so people shouldn't be surprised if they need to update their `additional-js`... Does that sound reasonable?

**Pull request merged**